File: src/main/java/mnm/mods/tabbychat/client/AbstractChannel.java
package mnm.mods.tabbychat.client;

import com.google.common.base.Strings;
import mnm.mods.tabbychat.api.Channel;
import net.minecraft.util.StringUtils;

public abstract class AbstractChannel implements Channel {

    private String alias;

    private String prefix = "";
    private boolean prefixHidden = false;

    private String command = "";

    public AbstractChannel(String name) {
        this.alias = name;
    }

    /**
     * Gets the alias that is displayed on the channel tab.
     *
     * @return The alias
     */
    public String getAlias() {
        return this.alias;
    }

    /**
     * Sets the alias that is displayed on the channel tab.
     *
     * @param alias The new alias
     */
    public void setAlias(String alias) {
        this.alias = alias;
    }

    /**
     * Gets the prefix that is inserted before the input.
     *
     * @return The prefix
     */
    public String getPrefix() {
        return Strings.nullToEmpty(this.prefix);
    }

    /**
     * Sets the prefix that is inserted before the input.
     *
     * @param prefix The new prefix
     */
    public void setPrefix(String prefix) {
        this.prefix = StringUtils.stripControlCodes(prefix);
    }

    /**
     * Gets whether the prefix is hidden or not. A hidden prefix will not be
     * displayed while typing. Defaults to false.
     *
     * @return True if the prefix is hidden
     */
    public boolean isPrefixHidden() {
        return this.prefixHidden;
    }

    /**
     * Sets whether the prefix is hidden or not. A hidden prefix will not be
     * displayed while typing.
     *
     * @param hidden Whether to hide the prefix
     */
    public void setPrefixHidden(boolean hidden) {
        this.prefixHidden = hidden;
    }

    public String getCommand() {
        return Strings.nullToEmpty(command);
    }

    public void setCommand(String command) {
        this.command = command;
    }

}

File: src/main/kotlin/mnm/mods/tabbychat/client/AbstractChannel.kt
package mnm.mods.tabbychat.client

import com.electronwill.nightconfig.core.Config
import mnm.mods.tabbychat.api.Channel
import net.minecraft.util.StringUtils

abstract class AbstractChannel(
        /**
         * The alias that is displayed on the channel tab.
         */
        open var alias: String = "",
        open var command: String = "",
        open var isPrefixHidden: Boolean = false,
        prefix: String = ""
) : Channel {

    /**
     * The prefix that is inserted before the input.
     */
    open var prefix: String = prefix
        set(value) {
            field = StringUtils.stripControlCodes(value)
        }

    fun toConfig(config: Config) {
        config.set<String>("name", name)
        config.set<String>("alias", alias)
        config.set<String>("command", command)
        config.set<Boolean>("isPrefixHidden", isPrefixHidden)
        config.set<String>("prefix", prefix)
    }

    companion object {
        fun <T : AbstractChannel> fromConfig(config: Config, factory: (String) -> T): T {
            return factory(config.get("name")).apply {
                alias = config.get("alias")
                command = config.get("command")
                isPrefixHidden = config.get("isPrefixHidden")
                prefix = config.get("prefix")
            }
        }
    }
}


